### DISCUSSION TITLE
How to extract Zod validation errors coming from TRPC?

### DISCUSSION THREAD

**Discussion by zivtamary** [trpc/trpc#3467] [Q&A]
Hi, I'm using the T3 stack, next.js/TRPC & Zod for input validation. whenever a validation fails in the router, TRPC throws a string of error object back to the client, and I'm unable extract the message from. 

when I try to log the err it's giving me this as string, not a real object.

 {
  message: '[\n' +
    '  {\n' +
    '    "code": "too_small",\n' 
    '    "minimum": 2,\n' 
    '    "type": "string",\n' 
    '    "inclusive": true,\n' 
    '    "exact": false,\n' 
    '    "message": "must be atleast 2 characters' 
    '    "path": [\n'
    '      "lastName"\n'
    '    ]\n' +
    '  }\n' +
    ']',
  code: -32600,
  data: {
    code: 'BAD_REQUEST',
    httpStatus: 400,
    stack: 'TRPCError: [\n' +
      '  {\n' +
      '    "code": "too_small",\n' +
      '    "minimum": 2,\n' +
      '    "type": "string",\n' +
      '    "inclusive": true,\n' +
      '    "exact": false,\n' +
      '    "message": "must be atleast 2 characters,\n' +
      '    "path": [\n' +
      '      "lastName"\n' +
      '    ]\n' +
      '  }\n' +
      ']\n' +
      '    at inputMiddleware (file:///D:/Courses/test/node_modules/@trpc/server/dist/index.mjs:252:19)\n' +
      '    at processTicksAndRejections (node:internal/process/task_queues:96:5)\n' +
      '    at async callRecursive (file:///D:/Courses/test/node_modules/@trpc/server/dist/index.mjs:419:32)\n' +
      '    at async next (file:///D:/Courses/test/node_modules/@trpc/server/dist/index.mjs:427:32)\n'' +
      '    at async callRecursive (file:///D:/Courses/test/node_modules/@trpc/server/dist/index.mjs:4419:32)\n' +
      '    at async resolve (file:///D:/Courses/test/node_modules/@trpc/server/dist/index.mjs:447:24))\n' +
      '    at async file:///D:/Courses/test/node_modules/@trpc/server/dist/resolveHTTPResponse-a3869dd43.mjs:123:32\n' +
      '    at async Promise.all (index 0)\n' +
      '    at async resolveHTTPResponse (file:///D:/Courses/test/node_modules/@trpc/server/dist/resollveHTTPResponse-a3869d43.mjs:120:28)\n' +
      '    at async nodeHTTPRequestHandler (file:///D:/Courses/test/node_modules/@trpc/server/dist/noodeHTTPRequestHandler-e46cee59.mjs:51:20)',
    path: 'user.updateProfile'
  }
}

**Comment by lpohren** [comment]
I'm having the same issue, @zivtamary have figured it out?

**Comment by KATT** [comment]
We have an example in the docs where we format zod errors: https://trpc.io/docs/server/error-formatting#adding-custom-formatting

**Reply by psanders** [reply]
This answer worked for me.

**Comment by lpohren** [comment]
@KATT even with that, I was having the issue.
I was able to fix it but using changing my node version.

At first I was using Node version `16.8.0`, which was leading to some install errors, which I was avoiding with `--ignore-engines`.
After I changed for Node `16.14.0` or `18.0.0`, I was able to install correctly without the ignore, and the Zod error was properly passed.
@zivtamary Try it out, might be you case as well.
